src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/loptimizer/heuristics/formula/FmAwareWeightedJoinAndUnboundVariableCount.java

```diff
@@ -0,0 +1,83 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.loptimizer.heuristics.formula;
+
+import org.apache.jena.graph.Node;
+import se.liu.ida.hefquin.engine.federation.FederationMember;
+import se.liu.ida.hefquin.engine.queryproc.impl.loptimizer.heuristics.utils.JoinAnalyzer;
+import se.liu.ida.hefquin.engine.queryproc.impl.loptimizer.heuristics.utils.QueryAnalyzer;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Estimate selectivity of a subplan by considering
+ * - 1. the number of new unbound variables
+ * - 2. the position of new unbound variables (sub, pred, or obj)
+ * - 3. number of joins contained in this subplan
+ * - 4. types of joins contained in this subplan (star join, chain join, unusual join)
+ *
+ * - 5. number of joins between this subplan with all selected plans
+ * - 6. type of joins between this subplan with all selected plans
+ *      - star join, chain join, or unusual join
+ *      - join within the same federation member or across different federation members
+ *
+ * This formula is an extended version of the formula that is implemented in {@link JoinAwareWeightedUnboundVariableCount}
+ */
+
+public class FmAwareWeightedJoinAndUnboundVariableCount extends JoinAwareWeightedUnboundVariableCount {
+
+    public static double estimate( final List<QueryAnalyzer> selectedPlans, final QueryAnalyzer nextPlan ) {
+        // Add bound variables of all selected plans to a set
+        final Set<Node> boundVariables = new HashSet<>();
+
+        for ( final QueryAnalyzer plan : selectedPlans ) {
+            boundVariables.addAll( plan.getSubs() );
+            boundVariables.addAll( plan.getPreds() );
+            boundVariables.addAll( plan.getObjs() );
+        }
+
+        return weightedUnboundVarsCount(boundVariables, nextPlan)
+                /
+                ( weightedJoinsCountInNextPlan( nextPlan )
+                        * weightedJoinsCountAcrossPlans( selectedPlans, nextPlan )
+                );
```

Comment:
I suggest to override the `calculateCost` function of `JoinAwareWeightedUnboundVariableCount`, move this part of the code into that overridden function, and then delete the `estimate` function from this class here.